packages/adblocker-extended-selectors/src/eval.ts

```diff
@@ -177,62 +181,96 @@ export function matches(element: Element, selector: AST): boolean {
   return false;
 }
 
+/**
+ * Describes CSS combinator behaviors from the given element.
+ * @param element The current subjective element.
+ * @param selector A complex selector.
+ */
 function handleComplexSelector(element: Element, selector: Complex): Element[] {
-  const elements: Element[] = [];
-  const leftElements =
+  // The *left* part of the given selector is not queried by the previous step.
+  // If there's no *left* part, we fallback to the current element.
+  const leftElements: Element[] =
     selector.left === undefined ? [element] : querySelectorAll(element, selector.left);
-
+  // The *right* part of the given selector is always *singular*.
+  // The understanding of `compound` selector behavior differs by `match` and `querySelectorAll`.
+  // The `compound` handler in `querySelectorAll` assume the subjective to be queried.
+  // However, our *actual* subjective elements are coming from *left* part of the selector.
+  // Therefore, we unmarshal the `compound` selector and directly use `traversal` which will involve `compound` handling in `match`.
+  const selectors =
+    selector.right.type === 'compound' ? selector.right.compound : [selector.right];
+  const results: Element[] = [];
   switch (selector.combinator) {
     case ' ':
-      for (const e of leftElements) {
-        elements.push(...querySelectorAll(e, selector.right));
+      // Look for all children *in any depth* of the all `leftElements` and filter them by `traversal`.
+      for (const leftElement of leftElements) {
+        for (const child of leftElement.querySelectorAll('*')) {
+          for (const result of traverse(child, selectors)) {
+            if (!results.includes(result)) {
+              results.push(result);
+            }
```

Comment:
instead of checking if results includes result, how about we make results a set?